loopd/daemon.go

```diff
@@ -119,117 +140,291 @@ func daemon(config *Config, lisCfg *listenerCfg) error {
+	d.swapClientServer = swapClientServer{
+		impl:        swapclient,
+		lnd:         &d.lnd.LndServices,
+		swaps:       make(map[lntypes.Hash]loop.SwapInfo),
+		subscribers: make(map[int]chan<- interface{}),
+		statusChan:  make(chan loop.SwapInfo),
+		mainCtx:     d.mainCtx,
+	}
+
+	// Retrieve all currently existing swaps from the database.
+	swapsList, err := d.impl.FetchSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swapsList {
+		d.swaps[s.SwapHash] = *s
+	}
+
+	// Start the main event handlers, now that we have prepared everything
+	// they need to work.
+	d.startSwapHandlers()
+
+	return nil
+}
 
+// startSwapHandlers starts all goroutines that are tasked with handling swap
+// events.
+func (d *Daemon) startSwapHandlers() {
 	// Start the swap client itself.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Starting swap client")
-		err := swapClient.Run(mainCtx, server.statusChan)
+		err := d.impl.Run(d.mainCtx, d.statusChan)
 		if err != nil {
 			log.Error(err)
+			d.internalErrChan <- err
 		}
 		log.Infof("Swap client stopped")
-
-		log.Infof("Stopping gRPC server")
-		grpcServer.Stop()
-
-		cancel()
 	}()
 
 	// Start a goroutine that broadcasts swap updates to clients.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Waiting for updates")
-		server.processStatusUpdates(mainCtx)
+		d.processStatusUpdates(d.mainCtx)
 	}()
 
-	// Start the grpc server.
-	wg.Add(1)
+	// Last, start our internal error handler. This will return at most one
+	// error on the main error channel to inform the caller that something
+	// went wrong.
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
-		log.Infof("RPC server listening on %s", grpcListener.Addr())
+		defer func() {
+			err := d.Stop()
+			if err != nil {
+				log.Errorf("Error stopping daemon: %v", err)
+			}
+			log.Info("Daemon exited")
+		}()
 
-		if restListener != nil {
-			log.Infof("REST proxy listening on %s", restListener.Addr())
-		}
+		select {
+		case err := <-d.internalErrChan:
+			log.Errorf("Runtime error in daemon, shutting down: "+
+				"%v", err)
+			d.mainErrChan <- err
 
-		err = grpcServer.Serve(grpcListener)
-		if err != nil {
-			log.Error(err)
+		case <-d.mainCtx.Done():
 		}
 	}()
+}
 
-	interruptChannel := make(chan os.Signal, 1)
-	signal.Notify(interruptChannel, os.Interrupt)
-
-	// Run until the users terminates loopd or an error occurred.
-	select {
-	case <-interruptChannel:
-		log.Infof("Received SIGINT (Ctrl+C).")
+// Stop tries to gracefully shut down the daemon.
+func (d *Daemon) Stop() error {
+	// If the swap client crashes, stop is called to make sure everything
+	// is cleaned up. But this also means we get here twice in case of the
+	// normal shutdown because after the swap client exists normally, stop
+	// is called again. We don't want to log an error in that case.
+	if atomic.AddInt32(&d.stopped, 1) != 1 {
+		return nil
```

Comment:
Isn't this a problem if a first stop is still in progress? It looks to the caller as if we already stopped.

Reply:
I don't see a scenario where this would be a problem. If a caller calls `Stop` just after the daemon crashed with an error then that error would be delivered over the error channel and handled there.

Reply:
Maybe they haven't read the error from the channel yet, because a ctrl-c was issued right before. The internal error handler is then still waiting for `Stop` to complete, but the ctrl-c stop call returns early here and may exit the process immediately, preventing clean shut down?

Reply:
That might be possible, yes. I added a wait on the wait group here. So the second call to `Stop` will return a nil error but only _after_ the first invocation is done.